torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,292 @@
+
+        :meth:`unscale` can be called explicitly to manipulate a network's
+        gradients after backward and before :meth:`step`.
+
+        Simple example, using :meth:`unscale` to enable clipping of unscaled gradients::
+
+            ...
+            scaler.scale(loss).backward()
+            scaler.unscale(optimizer)
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+            scaler.step(optimizer)
+            scaler.update()
+
+        If this instance of :class:`AmpScaler` is not enabled, :meth:`unscale` is a no-op.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that owns the gradients to be unscaled.
+
+        .. note::
+            :meth:`unscale` does not incur a CPU-GPU sync.
+
+        .. warning::
+            :meth:`unscale` should only be called once per optimizer per step,
+            and only after _all_ gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        rscale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device="cuda")
+
+        self._unscale_grads(optimizer, rscale, found_inf, allow_fp16=False)
+
+        optimizer_state["found_inf"] = found_inf
+        optimizer_state["unscaled"] = True
+
+    def step(self, optimizer, *args, **kwargs):
+        """
+        Carry out a scaling-safe step using ``optimizer``.  "Scaling-safe" means two things:
+
+        1.  If :meth:`unscale` has not yet been invoked, :meth:`step` will make sure to invoke :meth:`unscale`
+            internally.
+        2.  If inf/NaN gradients are found, :meth:`step` will skip ``optimizer.step()`` to avoid polluting the
+            params.
+
+        ``*args`` and ``**kwargs`` will be forwarded to ``optimizer.step()``.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that applies the gradients.
+            args:  Any arguments.
+            kwargs:  Any keyword arguments.
+
+        Returns:
+            The return value of ``optimizer.step(*args, **kwargs)``.
+
+        .. note::
+            If you're writing a custom optimizer, and wish to define your own scaling-safe ``step`` method that
+            :meth:`AmpScaler.step` may call directly without any wrapping logic, see the
+            :ref:`Custom Optimizer Guide<custom-optimizer-guide>`.
+
+        .. warning::
+            Closure use is not currently supported.
+        """
+        if (not self._enabled):
+            return optimizer.step(*args, **kwargs)
+
+        if "closure" in kwargs:
```

Comment:
Is this error still relevant?

Reply:
Yes, I'd like to at least detect it and bail out, so people know closures aren't doing what they think.